server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala

```diff
@@ -11,20 +12,18 @@ import scalaz.concurrent.Task
   * uri = "/foo/" to match the route.
   */
 object AutoSlash {
-
-  def apply(service: HttpService): HttpService = {
-    Service.lift { req =>
-      service.run(req).flatMap {
-        case r@Some(_) => Task.now(r)
-        case None      =>
-          val p = req.uri.path
-          if (p.isEmpty || p.charAt(p.length - 1) != '/') Task.now(None)
-          else {
-            val withSlash = req.copy(uri = req.uri.copy(path = p.substring(0, p.length - 1)))
-            service.run(withSlash)
-          }
-      }
+  def apply(service: HttpService): HttpService = { req =>
+    service(req).flatMap {
+      case resp if resp.status == Status.NotFound =>
```

Comment:
Where we used to key on a `None`, we now have to resort to `NotFound`.  This is a price we pay for the simplification.